File: server/internal/model/workout_set.go
package model

import (
	"time"

	"github.com/google/uuid"

	"wsTrack/server/internal/enum"
)

type WorkoutSet struct {
	ID                uuid.UUID       `gorm:"type:uuid;default:gen_random_uuid();primaryKey"`
	WorkoutExerciseID uuid.UUID       `gorm:"type:uuid;not null;index;uniqueIndex:idx_workout_sets_exercise_client"`
	WorkoutExercise   WorkoutExercise `gorm:"foreignKey:WorkoutExerciseID"`
	ClientID          string          `gorm:"size:64;not null;uniqueIndex:idx_workout_sets_exercise_client"`
	SetNumber         int             `gorm:"not null"`
	Weight            *float64
	Reps              *int
	DurationSeconds   *int
	Distance          *float64
	RPE               *float64 `gorm:"column:rpe"`
	IsWarmup          bool     `gorm:"default:false"`
	IsCompleted       bool     `gorm:"default:true"`
	RestSeconds       *int
	IsPR              bool            `gorm:"default:false"`
	Unit              enum.WeightUnit `gorm:"type:varchar(8);default:'kg'"`
	CompletedAt       *time.Time
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

File: mobile/src/models/WorkoutSet.ts
import { Model } from "@nozbe/watermelondb";
import type { Query, Relation } from "@nozbe/watermelondb";
import type { Associations } from "@nozbe/watermelondb/Model";
import { children, field, immutableRelation } from "@nozbe/watermelondb/decorators";

import type { WeightUnit } from "../constants/enums";
import type { PersonalRecord } from "./PersonalRecord";
import type { WorkoutExercise } from "./WorkoutExercise";
import type { WorkoutVideo } from "./WorkoutVideo";

export class WorkoutSet extends Model {
  static table = "workout_sets";

  static associations: Associations = {
    personal_records: { type: "has_many", foreignKey: "workout_set_id" },
    workout_exercises: { type: "belongs_to", key: "workout_exercise_id" },
    workout_videos: { type: "has_many", foreignKey: "workout_set_id" },
  };

  @field("workout_exercise_id") workoutExerciseId!: string;
  @field("set_number") setNumber!: number;
  @field("weight") weight!: number | null;
  @field("reps") reps!: number | null;
  @field("duration_seconds") durationSeconds!: number | null;
  @field("distance") distance!: number | null;
  @field("rpe") rpe!: number | null;
  @field("is_warmup") isWarmup!: boolean;
  @field("is_completed") isCompleted!: boolean;
  @field("rest_seconds") restSeconds!: number | null;
  @field("is_pr") isPr!: boolean;
  @field("unit") unit!: WeightUnit;
  @field("completed_at") completedAt!: number | null;
  @field("updated_at") updatedAt!: number;

  @immutableRelation("workout_exercises", "workout_exercise_id")
  workoutExercise!: Relation<WorkoutExercise>;
  @children("personal_records") personalRecords!: Query<PersonalRecord>;
  @children("workout_videos") videos!: Query<WorkoutVideo>;
}
